zcash_client_backend/src/data_api/chain.rs

```diff
@@ -341,7 +378,9 @@ where
     batch_runner.flush();
 
     let mut scanned_blocks = vec![];
-    block_source.with_blocks::<_, DbT::Error>(
+    let mut received_note_count = 0;
+    let mut spent_note_count = 0;
+    let scanned_range = block_source.with_blocks::<_, DbT::Error>(
```

Comment:
We're using the fact that `with_blocks` takes an `FnMut` to collect information about the blocks that the closure was called with. That means we could _also_ detect and collect `scanned_range` in the same way, without altering `BlockSource`. If we can't adequately explain what the return value of `with_blocks` is meant to be (per my previous comment), then an alternative is to not alter `BlockSource` and just collect the data in this context where we know what we want to report.

Reply:
Yeah, I thought about adding a `fold_blocks` method to `BlockSource` and then implementing `fold_blocks` in terms of it.